docs/proposals/scheduling/20230811-node-allocatable-resource-amplification.md

```diff
@@ -0,0 +1,88 @@
+  - "@zqzten"
+reviewers:
+  - "@saintube"
+  - "@hormes"
+creation-date: 2023-08-11
+last-updated: 2023-08-24
+---
+
+# Node Allocatable Resource Amplification
+
+## Table of Contents
+
+<!-- TOC -->
+* [Node Allocatable Resource Amplification](#node-allocatable-resource-amplification)
+  * [Table of Contents](#table-of-contents)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals/Future Work](#non-goalsfuture-work)
+  * [Proposal](#proposal)
+    * [User Stories](#user-stories)
+    * [Workflow](#workflow)
+    * [API](#api)
+    * [Constraints](#constraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+      * [Amplification Ratio Change](#amplification-ratio-change)
+      * [Unexpected Resource Overcommitment](#unexpected-resource-overcommitment)
+<!-- TOC -->
+
+## Summary
+
+Introduce a mechanism to amplify the allocatable resources of Kubernetes Nodes **at API level**, which enables the node to meet higher pod CPU requests. This feature can then be utilized by other higher-level ones such as CPU normalization.
+
+## Motivation
+
+There are cases where users would like to amplify the allocatable resources of a node, for example:
+
+- The node is equipped with a new-gen CPU, which is way faster than those on other nodes. In this case, users would like to amplify the allocatable CPU of the node in order to make most of the fast CPU of the node.
+- The node's allocatable CPU is nearly full but the actual CPU usage remains at an extremely low level. In this case, users would like to amplify the allocatable CPU of the node in order to improve the CPU utilization of the node as well as to increase the amount of schedulable pods of the cluster.
+
+### Goals
+
+- Introduce a mechanism to amplify allocatable resources of Kubernetes Nodes at API level.
+
+### Non-Goals/Future Work
+
+- Arrange underlying resources on Node with those amplified.
+
+## Proposal
+
+### User Stories
+
+This has been covered in [motivation](#motivation).
+
+### Workflow
+
+On k8s side, user (end user or other system) annotates the Node to specify its allocatable resource amplification ratio.
+
+On koord-manager side, the Node mutating webhook intercepts the Node status updating request from kubelet and modifies its allocatable CPU according to the amplification ratio. It also records the raw allocatable resources in annotations of the Node.
+
+### API
+
+- Introduce a new Node annotation `node.koordinator.sh/allocatable-amplification-ratio` whose value is a JSON object string of type `map[corev1.ResourceName]float32`. The float value must > 1. This is used by user to specify the amplification ratio for each allocatable resource of the Node.
+- Introduce a new Node annotation `node.koordinator.sh/raw-allocatable` whose value is a JSON object string of type `corev1.ResourceList`. This is used by koord-manager to record the Node's original allocatable resources (the `status.allocatable`), just informational for now.
```

Comment:
do we also amplify the `node.status.capacity`

Reply:
I prefer to keep the `node.status.capacity` as is since:
* The `node.status.capacity` is only informational (please point it out if I'm wrong), so it seems to be meaningless to amplify it.
* From the point of semantics, the `node.status.capacity` represent the actual physical resources of the Node, unlike `node.status.allocatable`, which could be a logical representation.

Reply:
I'm afraid that `node.status.capacity` is not informational. This field may not be consumed by the scheduler, but the calculated capacity value is used by the kubelet's container manager and other modules internally.

Reply:
As discussed, we need to scale the `capacity` to keep the assumption where `allocatable <= capacity`. However, we may not care about the capacity value inside the kubelet in the first version.